File: src/lib/api/github.ts
import { getTransport } from '$lib/api';
import type { GitHubAuthInfo, DeviceCodeResponse, GitHubUser, GitHubRepo } from './types';

export async function checkAuth(): Promise<GitHubAuthInfo> {
  return getTransport().invoke('github_check_auth');
}

export async function startDeviceFlow(): Promise<DeviceCodeResponse> {
  return getTransport().invoke('github_start_device_flow');
}

export async function pollDeviceFlow(
  deviceCode: string,
  interval: number,
  expiresIn: number
): Promise<GitHubAuthInfo> {
  return getTransport().invoke('github_poll_device_flow', { deviceCode, interval, expiresIn });
}

export async function logout(): Promise<void> {
  return getTransport().invoke('github_logout');
}

export async function getUser(): Promise<GitHubUser | null> {
  return getTransport().invoke('github_get_user');
}

export async function listRepos(): Promise<GitHubRepo[]> {
  return getTransport().invoke('github_list_repos');
}

File: src-tauri/src/commands/github.rs
use crate::github::auth;
use crate::github::{api, error::GitHubResult};
use crate::github::types::*;

#[tauri::command]
pub async fn github_check_auth() -> GitHubResult<GitHubAuthInfo> {
    auth::check_auth_status().await
}

#[tauri::command]
pub async fn github_start_device_flow() -> GitHubResult<DeviceCodeResponse> {
    auth::request_device_code().await
}

#[tauri::command]
pub async fn github_poll_device_flow(
    device_code: String,
    interval: u64,
    expires_in: u64,
) -> GitHubResult<GitHubAuthInfo> {
    auth::poll_for_token(&device_code, interval, expires_in).await
}

#[tauri::command]
pub async fn github_logout() -> GitHubResult<()> {
    auth::logout()
}

#[tauri::command]
pub async fn github_get_user() -> GitHubResult<Option<GitHubUser>> {
    let token = crate::github::credential::get_token();
    match token {
        Some(t) => {
            let user = auth::get_authenticated_user(&t).await?;
            Ok(Some(user))
        }
        None => Ok(None),
    }
}

#[tauri::command]
pub async fn github_list_repos() -> GitHubResult<Vec<GitHubRepo>> {
    api::list_user_repos().await
}
